odoo/models.py

```diff
@@ -3925,16 +3996,19 @@ def _create(self, vals):
                         pleft = cr.fetchone()[0] + 1
                 else:
                     # determine parent_left: it comes after all top-level parent_right
-                    cr.execute("SELECT MAX(parent_right) FROM %s" % self._table)
-                    pleft = (cr.fetchone()[0] or 0) + 1
-
-                # make some room for the new node, and insert it in the MPTT
-                cr.execute("UPDATE %s SET parent_left=parent_left+2 WHERE parent_left>=%%s" % self._table,
-                           (pleft,))
-                cr.execute("UPDATE %s SET parent_right=parent_right+2 WHERE parent_right>=%%s" % self._table,
-                           (pleft,))
-                cr.execute("UPDATE %s SET parent_left=%%s, parent_right=%%s WHERE id=%%s" % self._table,
-                           (pleft, pleft + 1, id_new))
+                    pleft = max_parent_right + 1
+
+                # make some room for the new node, and insert it in the MPTT;
+                # the nodes on the right are shifted in a way to avoid violating
+                # the uniqueness constraints on parent_left and parent_right:
+                # pos -> pos + M + 2 -> pos + 2
+                query = """ UPDATE {t} SET parent_left=parent_left+%(m)s+2 WHERE parent_left>=%(pleft)s;
```

Comment:
You can actually do something a bit more tricky, like

    UPDATE {t}
    SET parent_left=parent_left+(CASE WHEN parent_left>=%s THEN 2 ELSE 0 END),
        parent_right=parent_right+(CASE WHEN parent_right>=%s THEN 2 ELSE 0 END)
    WHERE parent_left>=%s OR parent_right>=%s

With this query, you update either `parent_left`, or `parent_right`, or both, depending on whether they are to the right of the new node.

I have tried this query, but it did not work, but the bug was not clear. I may give it another try...